core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/io/AwsCompressionInputStream.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.io;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.compression.Compressor;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.io.SdkInputStream;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * A wrapper class of InputStream that implements compression in chunks.
+ */
+@SdkInternalApi
+public class AwsCompressionInputStream extends SdkInputStream {
```

Comment:
I think we can use AwsChunkedEncodingInputStream instead of duplicating this class.
We can create an abstract classAwsChunkedInputStream with the common parts and reuse the code.

Reply:
got you, will refactor